interp/interp_test.go

```diff
@@ -191,3 +190,336 @@ func TestPiecewiseConstantPredict(t *testing.T) {
+		t.Errorf("Mismatch in coeffs: got %v, want %v", pc.coeffs, coeffs)
+	}
+}
+
+func TestPiecewiseCubicFitWithDerivativesErrors(t *testing.T) {
+	t.Parallel()
+	for _, test := range []struct {
+		xs, ys, dydxs []float64
+	}{
+		{
+			xs:    []float64{0, 1, 2},
+			ys:    []float64{10, 20},
+			dydxs: []float64{0, 0, 0},
+		},
+		{
+			xs:    []float64{0, 1, 1},
+			ys:    []float64{10, 20, 30},
+			dydxs: []float64{0, 0, 0, 0},
+		},
+		{
+			xs:    []float64{0},
+			ys:    []float64{0},
+			dydxs: []float64{0},
+		},
+		{
+			xs:    []float64{0, 1, 1},
+			ys:    []float64{10, 20, 10},
+			dydxs: []float64{0, 0, 0},
+		},
+	} {
+		var pc PiecewiseCubic
+		if !panics(func() { pc.fitWithDerivatives(test.xs, test.ys, test.dydxs) }) {
+			t.Errorf("expected panic for xs: %v, ys: %v and dydxs: %v", test.xs, test.ys, test.dydxs)
+		}
+	}
+}
+
+func TestAkimaSpline(t *testing.T) {
+	t.Parallel()
+	const (
+		nPts      = 40
+		wiggleTol = 1e-1
+	)
+	for i, test := range []struct {
+		xs  []float64
+		f   func(float64) float64
+		tol float64
+	}{
+		{
+			xs:  []float64{-1, 0, 1},
+			f:   func(x float64) float64 { return x * x },
+			tol: 1e-14,
+		},
+		{
+			xs:  []float64{-1, 1},
+			f:   func(x float64) float64 { return -10 * x },
+			tol: 1e-14,
+		},
+		{
+			xs:  []float64{-1, 0, 1},
+			f:   func(x float64) float64 { return -10 * x },
+			tol: 1e-14,
+		},
+		{
+			xs:  []float64{-0.2, -0.1, 0, 0.1, 0.2},
+			f:   math.Cos,
+			tol: 1e-4,
+		},
+		{
+			xs:  []float64{-0.2, -0.1, 0, 0.1, 0.2},
+			f:   math.Sin,
+			tol: 1e-4,
+		},
+		{
+			xs:  []float64{-0.2, -0.1, 0, 0.1, 0.2},
+			f:   math.Exp,
+			tol: 1e-4,
+		},
+		{
+			xs:  []float64{-0.2, -0.1, 0, 0.1, 0.2},
+			f:   func(x float64) float64 { return 1 / (1 + math.Exp(-100*x)) },
+			tol: 0.5,
+		},
+	} {
+		var as AkimaSpline
+		ys := applyFunc(test.xs, test.f)
+		err := as.Fit(test.xs, ys)
+		if err != nil {
+			t.Errorf("Error when fitting AkimaSpline in test case %d: %v", i, err)
+		}
+		x0 := test.xs[0]
+		x1 := test.xs[len(test.xs)-1]
+		dx := (x1 - x0) / nPts
+		for j := -1; j <= nPts+1; j++ {
+			x := x0 + float64(j)*dx
+			got := as.Predict(x)
+			want := test.f(math.Min(x1, math.Max(x0, x)))
+			if math.Abs(got-want) > test.tol {
+				t.Errorf("Mismatch in interpolated value at x == %g for test case %d: got %v, want %g", x, i, got, want)
+			}
+		}
+		n := len(test.xs)
+		for j := 0; j < n; j++ {
+			x := test.xs[j]
+			got := as.Predict(x)
+			want := test.f(x)
+			if math.Abs(got-want) > 1e-14 {
+				t.Errorf("Mismatch in interpolated value at x == %g for test case %d: got %v, want %g", x, i, got, want)
+			}
+		}
+		m := n - 1
+		for j := 0; j < m; j++ {
+			x0 := test.xs[j]
+			x1 := test.xs[j+1]
+			yMin := math.Min(ys[j], ys[j+1])
+			yMax := math.Max(ys[j], ys[j+1])
+			dx := (x1 - x0) / nPts
+			for k := 0; k <= nPts; k++ {
+				x := x0 + float64(k)*dx
+				y := as.Predict(x)
+				if y < yMin-wiggleTol || y > yMax+wiggleTol {
```

Comment:
[Kind of](https://en.wikipedia.org/wiki/Akima_spline): 

> Akima spline avoids unphysical wiggles in regions where the second derivative in the underlying curve is rapidly changing.

I tried to test it here. I think it's a useful test.

Reply:
I'm not saying it's not a useful test, I just don't think that "avoids unphysical wiggles" means "monotonicity of input implies monotonicity of the interpolant".

Can you try the following input to see if this test passes with it?
```
xs: []float64{7.99, 8.09, 8.19, 8.7, 9.2, 10, 12, 15, 20},
ys: []float64{0, 2.76429e-5, 4.37498e-2, 0.169183, 0.469428, 0.943740, 0.998636, 0.999919, 0.999994},
```

Reply:
I will. The test doesn't check whether the interpolant is monotonic, just that it doesn't shoot up or down too much.

Reply:
"too much" is relative and the current tolerance of 0.1 very loose and arbitrary. I haven't seen anything about Akima splines that would guarantee this property theoretically and so justify this test. If you had a spline that does provide theoretical guarantees, then this test with a tight tolerance should definitely stay.

Reply:
It's no less arbitrary than many of the tolerances in the codebase...

Reply:
Akima splines don't have the property that you're testing for, therefore this part of the test should be removed. To give another example input where this is demonstrated even better:
```
xs: []float64{1, 2, 4, 6.5, 8, 10, 10.5, 11, 13, 14},
ys: []float64{0, 0, 0, 0, 0.1, 1, 4.5, 8, 10, 15},
```
In `[8,10]` the test expects the interpolant to have values in `[0,1.1]. In reality, the spline goes down to roughly -1.3.

Not that it is relevant to this review but I'm not aware of any place in Gonum where we don't try to have the test tolerances as tight as possible. If we had to tweak them, it's either because the numerical solution is difficult for floating point arithmetic (e.g. in optimize, mat) or that we want the tests to finish in a reasonable time (in stat).

Reply:
Fair enough. I'll delete it.